### PR TITLE
[TRTLLM-14955][refactor] Declare MoE backend behaviour instead of comparing classes

### PR DESCRIPTION
**Pull request by xxi-nv**
## Description

`ExternalCommMoEScheduler` and `ConfigurableMoE` decided what to prepare for a backend by naming its class: 13 sites spread over `__class__ ==`, `isinstance`, and a five-way chain in `_get_backend_kwargs`. Adding a backend therefore meant editing the scheduler, and the checks disagreed with each other about subclasses — the multi-chunk LoRA guard was exact-class while the DWDP gate was `isinstance`.

Each site now reads a declaration off the backend class instead:

| Declaration | Replaces |
|---|---|
| `MoEStaticCapability` | multi-chunk LoRA guard, DWDP gate |
| `MoEInputRequirement` | routing-scale precision, `run_moe` workspace allocation, DeepEP expert-id sanitization, NVLink one-sided combine workspace dtype |
| `MoERunContext` + `MoECommPlan` | `run_moe`, which took 14 loosely-typed arguments assembled by the class chain |

Backends deriving from another backend restate every field rather than inherit it, because the old predicates were inconsistent. All 11 impl classes were verified to reproduce the old predicate for every declared field. That restatement rule is transitional and only has to hold while backends still derive from backends; TRTLLM-14960..14969 cut those inheritance edges.

**Two collapses are intentional.** The float32 assertion and the bfloat16 conversion become one cast, a no-op for the backends that used to assert. The sanitize request drops its `isinstance(moe.comm, DeepEP)` half because only `DeepEP.dispatch` reads the kwarg and every other strategy absorbs it through `**kwargs`.

**Removing the class chain also fixes latent defects**, because it had no `else` branch: any impl outside it silently received empty kwargs and fell back to signature defaults. On `CuteDslB12xFusedMoE` the CUTLASS prefill path was told `is_sf_swizzled=True` while `quantize_input` had produced unswizzled scale factors under post-quant dispatch, and `output_dtype` was forced to bfloat16 for fp16 models. `payload_in_workspace` is now assigned on every path, so it can no longer be inherited from a previous forward.

The five external-comm impls take the plan through `require_comm_plan()` rather than defaulting each field when it is absent. `comm_plan` stays optional on the context because a fused-comm impl owns the exchange, but an external-comm impl is only reachable through `ExternalCommMoEScheduler`, which builds a plan on every path. Defaulting there reintroduced the same class of defect in a new place: a wrong `moe_output` or `enable_alltoall` fails loudly, while a wrong `input_sf_swizzled` just makes the kernel read scale factors at the wrong stride. The three impls that had such a default disagreed on its value.

The `router_logits` filter that sat in the TRTLLMGen arm of the chain moves into `TRTLLMGenFusedMoE._routes_outside_the_kernel`, beside the kernel that cares about it; its three triggers are carried over unchanged. `FORCE_SEPARATED_ROUTING` moves to `interface.py` because both sides read it and that is the only module both already import.

No functional change is intended beyond the latent-defect fixes called out above.

## Test Coverage

This is a behaviour-preserving refactor, so the existing MoE suites are the safeguard. Both were exercised on GB300 (SM103, 4 GPU):

- `tests/unittest/_torch/modules/moe/test_moe_backend.py` — updated to build a `MoERunContext` and attach a `MoECommPlan` for external-comm backends, so every backend's `run_moe` is driven through the new contract. Covers Cutlass, CuteDSL, DeepGemm, DenseGEMM and TRTLLMGen across the quantization modes.
- `tests/unittest/_torch/modules/moe/test_moe_module.py` — end-to-end single-GPU and multi-GPU (EP/DP, EPLB) paths through `ConfigurableMoE` and both schedulers.
- `tests/unittest/_torch/lora/test_moe_lora_model_path.py` — updated for the `MoEStaticCapability.supports_moe_lora` declaration that replaces the exact-class LoRA check.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Refactors MoE backend behavior selection to use declared capabilities and input requirements.
- Introduces `MoERunContext` and `MoECommPlan` for typed execution and communication state.
- Updates MoE backends and wrappers to use the new `run_moe` API.
- Adds explicit handling for routing-scale dtype, expert-ID sanitization, workspace allocation, and output workspace support.
- Adds validation for required communication plans and runtime workspaces.
- No configuration or test-list files changed.
- No additional correctness issues were identified.

## QA Engineer Review

- Updates MoE backend tests for `MoERunContext`, `MoECommPlan`, workspace handling, and fused communication.
- Updates LoRA model-path tests for LoRA propagation and backend capability filtering.
- The modified test functions are not listed in `tests/integration/test_lists/`, `test-db/`, or `qa/` based on the available change summary.
- Verdict: needs follow-up.

### CODE REVIEW

File: tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py [hunk 1]
```diff
@@ -402,13 +390,15 @@ def _forward_chunk_impl(
...
-                assert token_final_scales.dtype == torch.float32
             assert token_selected_experts.dtype == torch.int32
 
-            if token_final_scales is not None and isinstance(moe.backend, TRTLLMGenFusedMoE):
-                token_final_scales = token_final_scales.to(torch.bfloat16)
+            # Backends disagree on routing-scale precision. The cast replaces
+            # both a float32 assertion and a bfloat16 conversion; for the
+            # backends that used to assert, routing already produces float32,
+            # so the cast is a no-op.
+            scales_dtype = moe.backend.input_requirement.routing_scales_dtype
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py:397]: Collapsing the float32 assertion into the cast also drops the invariant check: if a routing method ever stops producing float32, the cast silently fixes it instead of failing. Would it be worth keeping the `assert` alongside the cast, so an upstream contract break still surfaces?

File: tensorrt_llm/_torch/modules/fused_moe/interface.py [hunk 2]
```diff
@@ -927,32 +963,24 @@ def quantize_input(
     @abstractmethod
     def run_moe(
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/interface.py:964]: `MoEImplBase.run_moe` (impl_base.py:71) still declares `(self, ctx)` without the new `workspace` keyword, so the two copies of this contract now disagree. Since `requires_run_moe_workspace` backends need it, please align the abstract signature there too.

File: tensorrt_llm/_torch/modules/fused_moe/impl_contract.py [hunk 3]
```diff
@@ -85,8 +85,8 @@ class MoEInputRequirement:
     # MoEScheduler.
     requires_sanitized_expert_ids: bool = False
     requires_router_logits: bool = False
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/impl_contract.py]: `requires_router_logits` has no reader anywhere, and `MoECommPlan.payload_in_workspace` (line 256) is written but never read by any impl -- `combine()` reads the flag off the strategy, not off the plan. Either drop them or mark them explicitly as reserved for TRTLLM-14972, so the next reader does not assume they are live.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py [hunk 4]
```diff
@@ -1032,16 +1048,14 @@ def _run_moe_w4a16_nvfp4(
         tuner_num_tokens: Optional[int] = None,
         tuner_top_k: Optional[int] = None,
         moe_output: Optional[torch.Tensor] = None,
-        enable_alltoall: Optional[bool] = None,
+        enable_alltoall: bool = False,
```
Comment (zhaoyangwang-nvidia) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py]: `enable_alltoall` used to fall back to `self.enable_alltoall` when omitted; the new `bool = False` default silently means "no alltoall" instead, which selects a different expert-id remap branch. Consider making it a required keyword-only argument rather than defaulting it.

File: tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py (hunk 1 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/moe_scheduler.py:397]: Restored, keyed off the declared dtype instead of the class: a float32 request means the cast has to be a no-op, so the assert fires there, while TRTLLM-Gen's bfloat16 stays a deliberate narrowing.

Side effect is that the check now covers every backend rather than the two named classes. It also matters more than the old code implied: several routing methods take an `output_dtype` (`MiniMax2`, `SigmoidRenorm`, `Llama4`, `RenormalizeNaive`), and one configured narrower has already dropped bits that the cast would silently widen back.

File: tensorrt_llm/_torch/modules/fused_moe/interface.py (hunk 2 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/interface.py:964]: Deliberate, but you are right that nothing said so -- documented on both sides now.

`MoEImplBase` describes the state after impls allocate their own scratch through `get_workspaces` (the comment two lines below `run_moe`), so the parameter is missing on purpose rather than by omission. It also has no subclasses yet, so nothing currently has to satisfy both signatures. Adding `workspace` there would pull the transitional argument into the target contract, so instead each side states why it differs and when they converge (TRTLLM-14958, TRTLLM-14960..14969).

File: tensorrt_llm/_torch/modules/fused_moe/impl_contract.py (hunk 3 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/impl_contract.py]: Both confirmed, handled differently.

`requires_router_logits`: dropped. The design did intend it as the declarative replacement for the scheduler's router-logits filter, but that filter also reads the routing-method instance and an environment override, so a class-level bool cannot answer it -- `_routes_outside_the_kernel` does. Left a note where the field was so it does not come back from the design doc.

`payload_in_workspace`: kept, comment fixed. "combine() reads the same field" was wrong; `combine()` reads the same *value* off the strategy attribute, which the producer assigns from this field so the two cannot disagree. The comment now says that, plus the TRTLLM-14972 pointer.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py (hunk 4 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py]: Done, required keyword-only now.

Unreachable today (single caller, passes it explicitly), but the default is the wrong shape regardless: `False` selects the `- slot_start` remap, so a wrong guess shifts every expert id without failing.
